HTTP Multipart/FormData support

RX14 opened the pull request:
Lets try again!

asterite commented on src/http/common.cr:
```diff
@@ -223,6 +223,24 @@ module HTTP
     nil
   end
 
+  # Dequotes a RFC2616 quoted-string
```
Maybe show example code?

asterite commented on src/http/multipart/generator.cr:
```diff
@@ -0,0 +1,193 @@
+
+    # Finite State Machine diagram: https://gist.github.com/RX14/221c1edfa98d1196711515d4b5c264eb
+
+    # Generates a content type header with multipart subtype *subtype*, and
```
Maybe "Returns"? I initially thought this was appended to the generated content (because it says "generates" and this is a Generator class)

asterite commented on src/http/multipart/generator.cr:
```diff
@@ -0,0 +1,193 @@
+    end
+
+    # Yields an IO that can be used to write to a body part which is appended
+    # to the multipart message with the given *headers*. Throws is `#finish` or
```
"Throws **if**"

asterite commented on spec/std/http/multipart/parser_spec.cr:
```diff
@@ -0,0 +1,131 @@
+require "spec"
+require "http"
+
+def mp_parse(delim, data)
```
private def

RX14 replied:
This was written before those existed!

asterite commented on src/http/multipart/parser.cr:
```diff
@@ -0,0 +1,131 @@
+        begin
+          yield headers, body_io
+        ensure
+          # Read to end of body
```
I think here you can use `IO#skip_to_end` method that I added some days ago

(I'm happy to see more cases where this method is useful ^_^)

RX14 replied:
I think i'll send a PR which increases the buffer size of `skip_to_end`, `gets_to_end` so they trigger `IO::Buffered` to skip the buffer (>=4096).

asterite replied:
Good idea. I can commit that directly. What buffer size would be good?

RX14 replied:
I added this as a commit to this PR, you can cherry-pick it to master if you want.

asterite commented on src/http/formdata.cr:
```diff
@@ -0,0 +1,144 @@
+  # form_data = "--aA40\r\nContent-Disposition: form-data; name=\"field1\"\r\n\r\nfield data\r\n--aA40--"
+  # HTTP::FormData.parse(IO::Memory.new(form_data), "aA40") do |field, io|
+  #   field # => "field1"
+  #   data  # => "field data"
```
`io.gets_to_end # => "field data"`

asterite commented on src/http/formdata.cr:
```diff
@@ -0,0 +1,144 @@
+  #
+  # HTTP::FormData.parse(request) do |field, io|
+  #   field # => "field1"
+  #   data  # => "field data"
```
`io.gets_to_end # => "field data"`

asterite commented:
@RX14 Overall looks really good, I really like it! I don't know why I rejected it some time ago... sorry about that 😊 

I made a few comments for fixes. After that I'll merge it, so we can have it for 0.20.4. Thank you for this! 🎉

ysbaddaden commented on src/http/multipart.cr:
```diff
@@ -0,0 +1,81 @@
+  def self.parse(io, boundary)
+    parser = Parser.new(io, boundary)
+    while parser.has_next?
+      parser.next { |headers, io| yield headers, io }
```
Isn't it missing `meta` and `headers`? If not, how are we supposed to access them? Same below.

RX14 replied:
This is plain multipart, not multipart/form-data. Field names and file metadata are specific to the `multipart/form-data` mime-type.

ysbaddaden commented on src/http/multipart.cr:
```diff
@@ -0,0 +1,81 @@
+
+  # Yields a `Multipart::Generator` to the given block, returning the generated
+  # message as a `String`.
+  def self.generate(boundary = "--------------------------#{SecureRandom.urlsafe_base64(18)}")
```
What about extracting the default boundary generation to a helper method, to avoid repeating it over and over?

asterite commented:
I agree with all of @ysbaddaden comments. Let's postpone this PR for after 0.20.4 because I guess some changes are needed yet and I wanted to prepare a release today.

RX14 commented on src/http/multipart/parser.cr:
```diff
@@ -0,0 +1,131 @@
+  #
+  # Please note that the IO object yielded by `#next` is only valid until the
+  # block returns.
+  class Parser
```
That usecase is meant to be `HTTP::Multipart.parse`.

ysbaddaden replied:
But it's missing headers and meta?

RX14 commented:
I've found some more HTTP quoted-string craziness ([ref](https://stackoverflow.com/questions/7886782/what-is-the-exact-syntax-and-semantics-of-a-quoted-string-in-the-http1-1-rfc2616), [ref](https://evolvis.org/pipermail/evolvis-platfrm-discuss/2014-November/000675.html)), so i'll take the time to sort that out properly.

RX14 commented:
Updated the PR with my work today, github is fairly accurate on which points I have addressed and which I haven't. I'll split commits from the PR out into separate PRs on request.

RX14 commented:
There seems to be some confusion between `multipart/*` and `multipart/form-data`. RFC1341 defines that every media type using the `multipart` type must have a specific format. This format consists of a sequence of `body-part`s, each of which consists of any number of headers (including 0) and a body. It does not define any required headers.

The `multipart/form-data` subtype is an extension of the multipart media type which defines that each body part must contain a `Content-Disposition` header, which contains the name of the form field and optional metadata which is used when the field value is a file.

For that reason, `HTTP::Multipart` only yields headers and an IO, whereas `HTTP::FormData` yields extra data parsed from the `Content-Disposition` header. I'm probably going to create a `HTTP::FormData::Part` struct which contains the data from each part.

ysbaddaden commented:
Yes, I got confused. Thanks for the explanation.

Could you detail this in the docs for the different classes? Mostly insist on what they each apply to, and maybe note in `HTTP::Multipart` that we may be looking for `HTTP::FormData` instead?

In that light, I don't think the module helpers in `HTTP::Multipart` are needed. Having a few helpers in `HTTP::FormData` shall be enough. Other formats will use the `HTTP::Multipart` Parser and Builder classes directly.

Sija commented on src/http/formdata/parser.cr:
```diff
@@ -0,0 +1,70 @@
+
+  struct Part
+    getter name : String
+    getter 
```
looks like there's some code missing in here, I guess it's WIP :)

RX14 commented:
Sorry for the super-extended delay replying to these issue comments. I've done quite a bit of work on the PR, and it's ready to be reviewed again.

Sija commented on spec/std/http/http_spec.cr:
```diff
@@ -7,6 +7,11 @@ describe HTTP do
     HTTP.parse_time("Sun, 06 Nov 1994 08:49:37 GMT").should eq(time)
   end
 
+  it "parses RFC1123 without day name" do
```
space between `RFC` and its number would do nice ;)

RX14 replied:
Fixed.

RX14 commented:
I've rebased onto master so the CI should be fixed on OSX now. @ysbaddaden, another review would be appreciated :)

sdogruyol commented:
It'd be great to have this at next Crystal release please 🙏 👍

ysbaddaden commented:
@RX14 sorry, I've been quite busy lately, and haven't found the time to review this, yet.

asterite commented on src/http/common.cr:
````diff
@@ -229,6 +229,85 @@ module HTTP
+  # ```
+  def self.quote_string(string, io)
+    string.each_byte do |byte|
+      case QUOTE_LOOKUP_TABLE[byte]
````
We think it's better here to just do `case` on the byte, without a lookup table. The code will be clearer and it won't be unsafe (because it uses Pointer).

RX14 replied:
Will do. Do we know if LLVM will transform case statements to lookup tables?

RX14 replied:
Am I going crazy or is this a compiler error: https://gist.github.com/RX14/e2dd192c7cb0e7cfb2278c2c46f3bea3 ?

asterite replied:
The when takes the value of the comparison, so its comparing against a bool

asterite replied:
So it's like you are doing `if byte == (byte <= 0x1F)`

RX14 replied:
@asterite I could have sworn that i've seen case used like that before. I guess I shouldn't code at 1am.

asterite replied:
You can definitely do that if you omit the case expression. Then it becomes like if elsif but with all the conditions aligned.

RX14 replied:
@asterite committed the new version using `case`, and it's even 4% faster!

asterite commented on src/http/formdata/generator.cr:
````diff
@@ -0,0 +1,118 @@
+  # builder.finish
+  # io.to_s # => "--aA47\r\nContent-Disposition: form-data; name=\"name\"\r\n\r\njoe\r\n--aA47\r\nContent-Disposition: form-data; name=\"upload\"; filename=\"test.txt\"\r\n\r\nfile contents\r\n--aA47--"
+  # ```
+  class Builder
````
Could you rename the file (and the specs) to be named "builder" instead of "generator", now that the types are also named "Builder"?

RX14 replied:
Oh, sorry I thought I'd renamed them all! Thanks for the catch.

RX14 replied:
@asterite Renamed them

RX14 commented:
@ysbaddaden next time please can you leave a PR review instead of a commit comment, the commit comments get a bit lost when I force-push.

RX14 commented:
Interesting: this build failed due to lack of memory https://travis-ci.org/crystal-lang/crystal/jobs/201479637

RX14 commented:
Finally! :tada: :tada: :tada:
